### DISCUSSION TITLE
What causes HelmCharts to be re-reconciled?

### DISCUSSION THREAD

**Discussion by louismrose** [fluxcd/flux2#2555] [Q&A] [answered]
My Flux installation sometimes gets into an unusual state. I think the order of events is as follows:

1. Our GitRepo source in the cluster is re-reconciled.
2. Flux attempts to upgrade to a newer version of one of our Helm charts.
3. That Helm chart isn't yet in the repo's index.
4. The HelmChart resource in the cluster gets stuck in the state shown below. The corresponding HelmRelease resource never reconciles because its HelmChart isn't ready.
5. I manually resolve this by manually patching the HelmChart version on the HelmRelease to a non-existent version, and then manually patching back to the original version.
6. Flux reconciles everything and the HelmRelease upgrade completes successfully.

```
NAME                                 CHART                        VERSION   SOURCE KIND      SOURCE NAME     READY   STATUS                                                             AGE
flux-system-my-chart   my-chart   0.0.11    HelmRepository   helm-internal   False    Failed to find 'my-chart' chart with version '0.0.11' in helm-internal repository.   9d
```

NB: the "status" message above is an approximation of what I've seen; I haven't been able to reproduce this issue yet this morning.

My question is therefore -- what causes the `HelmChart` resource to re-reconcile? I can see that my `HelmChart` has an internal of 5 minutes. However, it doesn't look like the `HelmChart` is ever reconciled until I update the `HelmRelease` resource.

**Comment by kingdonb** [comment, marked as answer]
It sounds like this problem likely starts from a process that does "publish" and "release" in the same step. Can you reorder the steps so that "release" happens after "publish" is finished?

The HelmChart reconciles depending on the underlying source, which in this case I think is a GitRepository, but you haven't said that explicitly and I don't want to infer too much about what I think is wrong or how your pipeline is already structured.

Can you share more about your pipeline?

I think that you may have a new use for the new OCI features but I don't want to assume too much about your environment. Basically, the fact of keeping configuration blobs of any kind in an OCI artifact/image means that Flux's image update automation can now be used in places where it could not be used before. There are other improvements that make this change really significant, performance being top of mind, but back to your example.

I think that HelmRepository should eventually reconcile again and, seeing a new index.yaml, detect the chart that was missing before is now available. If it does not do this, then you may have found a bug. I'm not able to say for certain without more details about your setup. Here are some docs about manual reconciling of a HelmChart: https://fluxcd.io/docs/components/source/helmcharts/#working-with-helmcharts – rather than deleting the erroneous HelmChart (which should also be a way to mitigate the issue) you can annotate it with a timestamp, signaling to Flux that it needs a reconcile. HelmChart resources are not normally created by a user, and they do not normally change, so they don't have the same fixtures (eg. `flux reconcile source`) that other resources in Flux do get.

If Flux really doesn't recover from this condition even after all the involved objects `spec.interval` have passed, then I would consider this as a bug.

Everything in Flux reconciles on an interval, so when the HelmRepository is updated and the formerly missing version becomes available, I would expect that HelmRelease to reconcile a new HelmChart (or overwrite the old errored HelmChart) unless it's suspended. If it does not do this, please provide more detail so we have enough information to completely reproduce this.

But I would also suggest reorganizing your release pipeline somehow to avoid this condition. I would suggest separating the `Release chart` workflow from the `Publish release build (image)` workflow. If the image is not ready yet, then ostensibly it is because the chart is being published before the image has finished building – which leaves open the possibility that the image build will fail, and a chart is published referring to an image that will never be published.

We can test only so much when we build pipelines for CI/CD, and certain failures are unpredictable that may depend on stateful issues such as connection or service availability. The only way to guarantee against certain types of failure is to enforce the ordering so that a successful image being published definitely "happens before" the chart that refers to it gets promoted and released.

**Reply by kingdonb** [reply]
I think I have mixed some things up in my reply so please excuse those, I have confused the "publish reference to chart" step with the "publish release build (image)" step but I think the sentiment still applies. Please provide more information about your configuration if this answer doesn't fully solve your issue, and we can talk about the potential mitigations or solutions that Flux will offer soon.

I think that OCI feature (both the existing Helm OCI chart support, and the new Kustomize Controller OCI support that has yet to be released) is going to change an awful lot in a very good way, to make these sorts of issues less likely to naturally occur and more straightforward to work around. Being able to use Flux Image Automation in new and different places to connect CI releases with production deployments and include configuration without heavyweight index or git repository is a really substantial boon.

It already works today if you are using Helm, so if it's something that we can do to help you use this new feature in order to resolve the issue with your release workflow, I'd be really happy to be able to prove that with the help of real Flux users in public like this. 👍
